src/ucs/profile/profile.c

```diff
@@ -434,180 +433,211 @@ int ucs_profile_get_location(ucs_profile_type_t type, const char *name,
-            ctx->log.wraparound = 1;
+        if (++thread_ctx->log.current >= thread_ctx->log.end) {
+            thread_ctx->log.current    = thread_ctx->log.start;
+            thread_ctx->log.wraparound = 1;
         }
     }
 }
 
-static void ucs_profile_check_active_threads()
+static void ucs_profile_check_active_threads(ucs_profile_context_t *ctx)
 {
     size_t num_active_threads;
 
-    pthread_mutex_lock(&ucs_profile_global_ctx.mutex);
-    num_active_threads = ucs_list_length(&ucs_profile_global_ctx.thread_list);
-    pthread_mutex_unlock(&ucs_profile_global_ctx.mutex);
+    pthread_mutex_lock(&ctx->mutex);
+    num_active_threads = ucs_list_length(&ctx->thread_list);
+    pthread_mutex_unlock(&ctx->mutex);
 
     if (num_active_threads > 0) {
         ucs_warn("%zd profiled threads are still running", num_active_threads);
     }
 }
 
-void ucs_profile_reset_locations()
+void ucs_profile_reset_locations(ucs_profile_context_t *ctx)
 {
     ucs_profile_global_location_t *loc;
 
-    pthread_mutex_lock(&ucs_profile_global_ctx.mutex);
+    pthread_mutex_lock(&ctx->mutex);
 
-    ucs_profile_for_each_location(loc) {
+    ucs_profile_ctx_for_each_location(ctx, loc) {
         *loc->loc_id_p = -1;
     }
 
-    ucs_profile_global_ctx.num_locations = 0;
-    ucs_profile_global_ctx.max_locations = 0;
-    ucs_free(ucs_profile_global_ctx.locations);
-    ucs_profile_global_ctx.locations = NULL;
+    ctx->num_locations = 0;
+    ctx->max_locations = 0;
+    ucs_free(ctx->locations);
+    ctx->locations = NULL;
 
-    pthread_mutex_unlock(&ucs_profile_global_ctx.mutex);
+    pthread_mutex_unlock(&ctx->mutex);
 }
 
-static void ucs_profile_cleanup_completed_threads()
+static void ucs_profile_cleanup_completed_threads(ucs_profile_context_t *ctx)
 {
-    ucs_profile_thread_context_t *ctx, *tmp;
+    ucs_profile_thread_context_t *thread_ctx, *tmp;
 
-    pthread_mutex_lock(&ucs_profile_global_ctx.mutex);
-    ucs_list_for_each_safe(ctx, tmp, &ucs_profile_global_ctx.thread_list,
-                           list) {
-        if (ctx->is_completed) {
-            ucs_profile_thread_cleanup(ctx);
+    pthread_mutex_lock(&ctx->mutex);
+    ucs_list_for_each_safe(thread_ctx, tmp, &ctx->thread_list, list) {
+        if (thread_ctx->is_completed) {
+            ucs_profile_thread_cleanup(ctx->profile_mode, thread_ctx);
         }
     }
-    pthread_mutex_unlock(&ucs_profile_global_ctx.mutex);
+    pthread_mutex_unlock(&ctx->mutex);
 }
 
-void ucs_profile_dump()
+void ucs_profile_dump(ucs_profile_context_h ctx)
 {
-    ucs_profile_thread_context_t *ctx;
+    ucs_profile_thread_context_t *thread_ctx;
 
     /* finalize profiling on current thread */
-    ctx = pthread_getspecific(ucs_profile_global_ctx.tls_key);
-    if (ctx) {
-        ucs_profile_thread_finalize(ctx);
-        pthread_setspecific(ucs_profile_global_ctx.tls_key, NULL);
+    thread_ctx = pthread_getspecific(ctx->tls_key);
+    if (thread_ctx) {
```

Comment:
minor: since you already changing it, can you please set `thread_ctx != NULL`, according to code style?